### PR TITLE
breaking: require framework_version, py_version for tensorflow

### PR DESCRIPTION
**Pull request by metrizable**
*Issue #, if available:* #1465

*Description of changes:*

changes include:

* framework_version, py_version required for framework `TensorFlow`
* framework_version, py_version required for framework `TensorFlowModel`
* re-order of args: convention to follow entry_point, push args with defaults below
* testing updates
* doc updates

*Testing done:*

unit:
```
% tox --parallel 3 tests/unit
✔ OK black-format in 13.57 seconds
✔ OK flake8 in 24.793 seconds
✔ OK twine in 14.049 seconds
✔ OK pylint in 37.106 seconds
✔ OK doc8 in 19.171 seconds
✔ OK py36 in 41.295 seconds
✔ OK sphinx in 1 minute, 37.846 seconds
✔ OK py37 in 37.777 seconds
✔ OK py27 in 1 minute, 55.877 seconds
```
integs:
```
% export IGNORE_COVERAGE=- ; tox -e py37 -- -s -vv tests/integ/test_tuner.py -k _tf ; unset IGNORE_COVERAGE
...
tests/integ/test_tuner.py::test_tuning_tf_script_mode[1.15.2] Started hyperparameter tuning job with name: tune-tf-script-m-1591919001-d74c
....................................................!
PASSED
tests/integ/test_tuner.py::test_tuning_tf[1.15.2] SKIPPED
tests/integ/test_tuner.py::test_tuning_tf_vpc_multi[1.15.2] SKIPPED
tests/integ/test_tuner.py::test_tuning_tf_script_mode[2.2.0] Started hyperparameter tuning job with name: tune-tf-script-m-1591919286-6da1
...............................................................!
PASSED
...
```

```
% export IGNORE_COVERAGE=- ; tox -e py37 -- -s -vv tests/integ/test_tf.py -k "not ml.p2" ; unset IGNORE_COVERAGE
...
tests/integ/test_tf.py::test_mnist_with_checkpoint_config[1.15.2-ml.m4.xlarge]
...
2020-06-12 00:07:26 Completed - Training job completed
Training seconds: 134
Billable seconds: 134
PASSED
tests/integ/test_tf.py::test_mnist_distributed[1.15.2-ml.m4.xlarge]
...
2020-06-12 00:13:56 Completed - Training job completed
Training seconds: 386
Billable seconds: 386
PASSED
...

```
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to any/all clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/unit/sagemaker/tensorflow/test_estimator.py [hunk 1]
```diff
@@ -21,8 +21,9 @@
 import pytest
 
 from sagemaker.estimator import _TrainingJob
-from sagemaker.tensorflow import defaults, model, TensorFlow
-from tests.unit import DATA_DIR
+from sagemaker.tensorflow import model, TensorFlow
+from sagemaker.tensorflow.defaults import TF_VERSION
+from tests.unit import DATA_DIR, PY_VERSION
```
Comment (laurenyu) [inline tests/unit/sagemaker/tensorflow/test_estimator.py:26]: iirc `TF_VERSION` is `"1.10"` and `PY_VERSION` is `"py3"`, which isn't actually a valid combination if you tried it in an integ test. I'm guessing it doesn't matter in these tests, though.

File: doc/frameworks/tensorflow/upgrade_from_legacy.rst [hunk 2]
```diff
@@ -104,6 +104,7 @@ the difference in code would be as follows:
         ...
         source_dir="code",
         framework_version="1.10.0",
+        py_version="py3",
```
[the remark this answers is not in the document]
Comment (metrizable) [reply doc/frameworks/tensorflow/upgrade_from_legacy.rst:107]: updating

File: src/sagemaker/tensorflow/estimator.py [hunk 3]
```diff
@@ -56,11 +56,12 @@ def __init__(
 
         Args:
             py_version (str): Python version you want to use for executing your model training
-                code (default: 'py2').
+                code. One of 'py2', 'py3', or 'py37'. Defaults to ``None``. Required unless
```
[the remark this answers is not in the document]
Comment (metrizable) [reply src/sagemaker/tensorflow/estimator.py:59]: will do

File: src/sagemaker/tensorflow/estimator.py [hunk 4]
```diff
@@ -150,7 +149,7 @@ def __init__(
     def _validate_args(self, py_version, framework_version):
         """Placeholder docstring"""
 
-        if py_version == "py3":
+        if py_version:
             if framework_version is None:
                 raise AttributeError(fw.EMPTY_FRAMEWORK_VERSION_ERROR)
```
[the remark this answers is not in the document]
Comment (metrizable) [reply src/sagemaker/tensorflow/estimator.py:154]: absolutely. need to remove

File: src/sagemaker/tensorflow/model.py [hunk 5]
```diff
@@ -138,7 +137,8 @@ def __init__(
         role,
         entry_point=None,
         image=None,
-        framework_version=TF_VERSION,
+        framework_version=None,
+        py_version=None,
```
[the remark this answers is not in the document]
Comment (metrizable) [reply src/sagemaker/tensorflow/model.py:141]: sounds good

File: src/sagemaker/tensorflow/estimator.py [hunk 6]
```diff
@@ -114,18 +119,13 @@ def __init__(
...
         """
-        if framework_version is None:
-            logger.warning(
-                fw.empty_framework_version_warning(defaults.TF_VERSION, self.LATEST_VERSION)
-            )
-        self.framework_version = framework_version or defaults.TF_VERSION
-
-        if not py_version:
-            py_version = "py3" if self._only_python_3_supported() else "py2"
+        fw.validate_version_or_image_args(framework_version, py_version, image_name)
```
[the remark this answers is not in the document]
Comment (metrizable) [reply src/sagemaker/tensorflow/estimator.py:122]: agreed.

File: tests/integ/test_tf.py [hunk 7]
```diff
@@ -19,12 +19,11 @@
 import pytest
 
 from sagemaker.tensorflow import TensorFlow
-from sagemaker.tensorflow.defaults import LATEST_SERVING_VERSION
+from sagemaker.tensorflow.defaults import LATEST_VERSION, LATEST_SERVING_VERSION
```
[the remark this answers is not in the document]
Comment (metrizable) [reply tests/integ/test_tf.py:22]: will remove

File: tests/unit/__init__.py [hunk 8]
```diff
@@ -39,3 +39,5 @@
     "cn-northwest-1",
     "us-gov-west-1",
 ]
+
+PY_VERSION = "py3"
```
[the remark this answers is not in the document]
Comment (metrizable) [reply tests/unit/__init__.py:43]: agreed. sounds good

File: tests/conftest.py [hunk 9]
```diff
@@ -226,6 +225,16 @@ def tf_version(request):
     return request.param
 
 
+@pytest.fixture(scope="module", params=["py2", "py3"])
+def tf_py_version(tf_version, request):
+    version = [int(val) for val in tf_version.split(".")]
+    if version < [1, 13]:
+        return "py2"
```
Comment (laurenyu) [inline tests/conftest.py:232]: 1.11-1.12 have Python 3 support (those are the two versions with both legacy and current TF support)

File: tests/conftest.py [hunk 10]
```diff
@@ -209,7 +209,6 @@ def xgboost_version(request):
         "1.11.0",
         "1.12",
         "1.12.0",
-        "1.13",
```
Comment (laurenyu) [inline tests/conftest.py:212]: what's the purpose behind removing this?

File: tests/conftest.py [hunk 11]
```diff
@@ -290,6 +299,23 @@ def tf_full_version(request):
...
+def tf_full_py_version(tf_full_version, request):
+    """fixture to match tf_full_version
+
+    Fixture exists as such, since tf_full_version may be overridden --tf-full-version.
+    Otherwise, this would simply be py37 to match the latest version support.
+
+    TODO: Evaluate use of --tf-full-version with possible eye to remove and simplify code.
+    """
+    version = [int(val) for val in tf_full_version.split(".")]
+    if version < [1, 13]:
```
Comment (laurenyu) [inline tests/conftest.py:312]: same here

```suggestion
    if version < [1, 11]:
```

File: tests/integ/test_tuner.py [hunk 12]
```diff
@@ -639,6 +633,7 @@ def test_tuning_tf_script_mode(sagemaker_session, cpu_instance_type, tf_full_ver
         tuner.wait()
 
 
+# TODO: evaluate skip mark and default framework_version 1.11
```
Comment (laurenyu) [inline tests/integ/test_tuner.py:636]: we can delete this test - looks like I forgot about it when I was cleaning up TF

File: tests/integ/test_tuner.py [hunk 13]
```diff
@@ -693,6 +690,7 @@ def test_tuning_tf(sagemaker_session, cpu_instance_type):
         assert dict_result == list_result
 
 
+# TODO: evaluate skip mark and default framework_version 1.11
```
Comment (laurenyu) [inline tests/integ/test_tuner.py:693]: I'll take care of this in #1461

File: tests/conftest.py (hunk 9 above)
Comment (metrizable) [reply tests/conftest.py:232]: scriptmode, thank you.

File: tests/conftest.py (hunk 10 above)
Comment (metrizable) [reply tests/conftest.py:212]: ah. i see that it's still in DLC ECR repo, but just not listed on the supported versions:

* https://aws.amazon.com/releasenotes/available-deep-learning-containers-images/

i'll add this back in

File: tests/conftest.py (hunk 11 above)
Comment (metrizable) [reply tests/conftest.py:312]: updating

**Pull request merged**